### PR TITLE
Fixes

### PR DESCRIPTION
**Pull request by martijn00**
Also fixes https://github.com/minio/minio-dotnet/issues/772

@ebozduman

### CODE REVIEW

**Comment by martijn00** [comment]
@ebozduman can you merge?

**Comment by ebozduman** [comment]
@martijn00 

I still see the same problem when `CopyObject_Test8` functional test runs against framework `net4.8`.
Your fix looks complete, but still the same error:
```
{"name":"minio-dotnet : CopyObject_Test8","function":"Task<CopyObjectResult> CopyObjectAsync(CopyObjectArgs args, CancellationToken cancellationToken = default(CancellationToken))","description":"Tests whether CopyObject with metadata replacement passes","args":{"bucketName":"minio-dotnet-example-ehym8w6609mh64g","objectName":"bf@ghe+#h+","destBucketName":"minio-dotnet-example-fwoxevgbqvmaec6","destObjectName":"minio-dotnet-example-hfko4gkvpn","data":"1KB","size":"1KB","copyconditions":"x-amz-metadata-directive:REPLACE"},"duration":1051,"status":"FAIL","alert":"Collection was modified; enumeration operation may not execute.","message":"System.InvalidOperationException: Collection was modified; enumeration operation may not execute.\n  at System.Collections.Generic.Dictionary`2+Enumerator[TKey,TValue].MoveNext () [0x00013] in <de882a77e7c14f8ba5d298093dde82b2>:0 \n  at Minio.CopyObjectArgs.Populate () [0x000fc] in <ae63953852d443a790f9788c050ef577>:0 \n  at Minio.CopyObjectArgs.Validate () [0x00093] in <ae63953852d443a790f9788c050ef577>:0 \n  at Minio.MinioClient.CopyObjectAsync (Minio.CopyObjectArgs args, System.Threading.CancellationToken cancellationToken) [0x0023f] in <ae63953852d443a790f9788c050ef577>:0 \n  at Minio.Functional.Tests.FunctionalTest.CopyObject_Test8 (Minio.MinioClient minio) [0x00575] in <87b15959cb13430c973a5940771c37d1>:0 "}

Unhandled Exception:
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
  at System.Collections.Generic.Dictionary`2+Enumerator[TKey,TValue].MoveNext () [0x00013] in <de882a77e7c14f8ba5d298093dde82b2>:0 
  at Minio.CopyObjectArgs.Populate () [0x000fc] in <ae63953852d443a790f9788c050ef577>:0 
  at Minio.CopyObjectArgs.Validate () [0x00093] in <ae63953852d443a790f9788c050ef577>:0 
  at Minio.MinioClient.CopyObjectAsync (Minio.CopyObjectArgs args, System.Threading.CancellationToken cancellationToken) [0x0023f] in <ae63953852d443a790f9788c050ef577>:0 
  at Minio.Functional.Tests.FunctionalTest.CopyObject_Test8 (Minio.MinioClient minio) [0x0077f] in <87b15959cb13430c973a5940771c37d1>:0 
  at Minio.Functional.Tests.FunctionalTest.CopyObject_Test8 (Minio.MinioClient minio) [0x008a2] in <87b15959cb13430c973a5940771c37d1>:0 
  at Minio.Functional.Tests.Program.Main (System.String[] args) [0x004db] in <87b15959cb13430c973a5940771c37d1>:0 
  at Minio.Functional.Tests.Program.<Main> (System.String[] args) [0x0000c] in <87b15959cb13430c973a5940771c37d1>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
  at System.Collections.Generic.Dictionary`2+Enumerator[TKey,TValue].MoveNext () [0x00013] in <de882a77e7c14f8ba5d298093dde82b2>:0 
  at Minio.CopyObjectArgs.Populate () [0x000fc] in <ae63953852d443a790f9788c050ef577>:0 
  at Minio.CopyObjectArgs.Validate () [0x00093] in <ae63953852d443a790f9788c050ef577>:0 
  at Minio.MinioClient.CopyObjectAsync (Minio.CopyObjectArgs args, System.Threading.CancellationToken cancellationToken) [0x0023f] in <ae63953852d443a790f9788c050ef577>:0 
  at Minio.Functional.Tests.FunctionalTest.CopyObject_Test8 (Minio.MinioClient minio) [0x0077f] in <87b15959cb13430c973a5940771c37d1>:0 
  at Minio.Functional.Tests.FunctionalTest.CopyObject_Test8 (Minio.MinioClient minio) [0x008a2] in <87b15959cb13430c973a5940771c37d1>:0 
  at Minio.Functional.Tests.Program.Main (System.String[] args) [0x004db] in <87b15959cb13430c973a5940771c37d1>:0 
  at Minio.Functional.Tests.Program.<Main> (System.String[] args) [0x0000c] in <87b15959cb13430c973a5940771c37d1>:0 
```

**Comment by martijn00** [comment]
@ebozduman are you sure you pulled in the latest code from this branch? I explicetly test this on net4.8. I run CopyObject_Test8 again just now on 4.8 and it works fine now. Is there something I miss here?

File: Minio/DataModel/AccessCredentials.cs [hunk 1]
```diff
@@ -26,9 +26,14 @@ public class AccessCredentials
...
-            throw new ArgumentNullException(nameof(accessKey) + " and " + nameof(secretKey) +
-                                            " cannot be null or empty.");
+        if (string.IsNullOrEmpty(accessKey))
+            throw new ArgumentException($"'{nameof(accessKey)}' cannot be null or empty.", nameof(accessKey));
+
+        if (string.IsNullOrEmpty(secretKey))
+            throw new ArgumentException($"'{nameof(secretKey)}' cannot be null or empty.", nameof(secretKey));
+
+        if (string.IsNullOrEmpty(sessionToken))
+            throw new ArgumentException($"'{nameof(sessionToken)}' cannot be null or empty.", nameof(sessionToken));
```
Comment (ebozduman) [inline Minio/DataModel/AccessCredentials.cs:36]: I think there is a logic mistake here.
We don't need all 3 parameters to be provided all together, i.e. either `accessKey` and `secretKey` or `sessionToken` and `Expiration` need to be provided. 

So, any one of the pair is good for access to be granted. Of course at least one set should to be required.
I'll also double check Go SDK to make sure it is implemented this way.

**Review by ebozduman — CHANGES_REQUESTED** [review]

**Comment by martijn00** [comment]
@ebozduman I've fixed the remark. Can you run CI again and merge?

File: Minio/DataModel/AccessCredentials.cs [hunk 2]
```diff
@@ -26,19 +26,24 @@ public class AccessCredentials
...
+        if (!string.IsNullOrEmpty(accessKey) || !string.IsNullOrEmpty(secretKey) || !string.IsNullOrEmpty(sessionToken))
+        {
+            AccessKey = accessKey;
+            SecretKey = secretKey;
+            SessionToken = sessionToken;
+            Expiration = expiration.Equals(default) ? null : Utils.To8601String(expiration);
+        }
+        else
+        {
+            if (string.IsNullOrEmpty(secretKey))
```
Comment (ebozduman) [inline Minio/DataModel/AccessCredentials.cs:38]: `secretKey` => `accessKey`

**Review by ebozduman — CHANGES_REQUESTED** [review]

**Comment by martijn00** [comment]
@ebozduman good find. Fixed now.

**Review by ebozduman — APPROVED** [review]
LGTM

**Pull request merged**